### PR TITLE
Async stack trace structure (from dev)

### PR DESCRIPTION
**Pull request by clementbera**
Version rebased on dev branch for SMarr review.

So far basic synchronous on asynchronous benchmark are working with the different settings enabled. With the settings disabled everything should keep working too. Still need to work on printing the stack with the different settings enabled or not.

Three additional settings:
--async_stack_traces_structure: enables the shadow stack entry, allowing to recreate the stack across asynchronous calls.
--async_stack_traces_method_cache: enables per method cache to CachedDispatchNode or LexicallyBoundDispatchNode, avoiding to create shadow stack entry if a method is called from a single dispatch node.
--async_stack_traces_inline_cache: enables per shadow stack entry creation site cache, avoiding to create repeatedly the same shadow stack entry in loops, the first few times a loop is created.

async_stack_traces_structure has to be enabled for the other ones to change anything. The other 2 settings can be enabled separately or together.

### CODE REVIEW

File: build.xml [hunk 1]
```diff
@@ -21,11 +21,11 @@
     <property name="svm.build"   location="${svm.dir}/mxbuild/dists" />
     <property name="truffle.build" location="${truffle.dir}/mxbuild/dists" />
     <property name="somns-deps.version" value="0.3.6" />
-    <property name="checkstyle.version" value="8.11" />
+    <property name="checkstyle_checkse.version" value="8.11" />
```
Comment (smarr) [inline build.xml]: Why this change?

File: docs/basic-setup.md [hunk 2]
```diff
@@ -48,4 +48,4 @@ Afterwards, the simple Hello World program is executed with:
 
 ```bash
 ./som core-lib/Hello.ns
-```
+```
```
Comment (smarr) [inline docs/basic-setup.md]: Strange change?

File: src/som/interpreter/Invokable.java [hunk 3]
```diff
@@ -46,6 +46,10 @@ public final boolean isAtomic() {
     return isAtomic;
   }
 
+  public ExpressionNode getRoot() {
```
Comment (smarr) [inline src/som/interpreter/Invokable.java]: `Node.getRootNode()` isn't any good for this?

File: src/som/interpreter/Method.java [hunk 4]
```diff
@@ -77,6 +80,20 @@ public boolean equals(final Object o) {
     return false;
   }
 
+  public void setNewCaller(final BackCacheCallNode caller) {
```
Comment (smarr) [inline src/som/interpreter/Method.java:82]: You are changing a `@CompilationFinal` field here, do you need to invalidate any code that may have been compiled with this assumption backed in?

File: src/som/interpreter/actors/ErrorPromiseNode.java [hunk 5]
```diff
@@ -10,21 +10,23 @@
 
 
 @GenerateNodeFactory
-@Primitive(primitive = "actorsError:with:isBPResolver:isBPResolution:")
+@Primitive(primitive = "actorsError:with:entry:isBPResolver:isBPResolution:")
+
```
Comment (smarr) [inline src/som/interpreter/actors/ErrorPromiseNode.java]: Why the space?
Also, you seem to change the signature of the specialization, I'd assume that primitive signature would need to change then, too?

**Review by smarr — COMMENTED** [review]
Some very initial comments, only skimming over the first bits

File: build.xml (hunk 1 above)
Comment (clementbera) [reply build.xml]: I think this was generated by mx.

File: src/som/interpreter/Invokable.java (hunk 3 above)
Comment (clementbera) [reply src/som/interpreter/Invokable.java]: I think Node.getRootNode answers the root node, i.e. the method or primitive, while this method  answers the sequence of expression held by the root node. So I would say no.

File: docs/basic-setup.md (hunk 2 above)
Comment (clementbera) [reply docs/basic-setup.md]: Yes, the change should be removed but I don't know how. I checked out from dev and it is still there.

File: src/som/interpreter/Method.java (hunk 4 above)
Comment (clementbera) [reply src/som/interpreter/Method.java:82]: Oh, I thought @CompilationFinal was taking care of that for me. Yes code needs to be invalidated.

File: src/som/interpreter/actors/ErrorPromiseNode.java (hunk 5 above)
Comment (clementbera) [reply src/som/interpreter/actors/ErrorPromiseNode.java]: I removed the space.

I am not sure about what you are asking for the signature. The signature of AbstractPromiseResolutionNode was also changed, is that what you mean?

File: src/som/interpreter/Method.java (hunk 4 above)
Comment (clementbera) [reply src/som/interpreter/Method.java:82]: So I removed the @CompilationFinal of this variable. The problem is in BackCacheCallNode >> multipleCaller (). When this method is called, all the optimized code using the BackCacheCallNode has to be invalidated. I am not sure on how to do that.

File: build.xml (hunk 1 above)
Comment (smarr) [reply build.xml]: No, please revert. mx doesn't know anything about our build system.

File: src/som/interpreter/Invokable.java (hunk 3 above)
Comment (smarr) [reply src/som/interpreter/Invokable.java]: would `getBodyNode()` be an appropriate name?

File: src/som/interpreter/Method.java (hunk 4 above)
Comment (smarr) [reply src/som/interpreter/Method.java:82]: Not sure this is the ideal example: https://github.com/smarr/SOMns/blob/dev/src/som/interpreter/objectstorage/ObjectLayout.java#L153

Assumptions are the way to invalidate code "non-locally". Any code that is compiled checking an assumption is registered for being discarded when the assumption is invalidated. In the example, there is an assumption for the memory layout of classes. It is invalidated, when the class gets a new layout assigned (which in return is going to be used by its instances).

**Comment by smarr** [comment]
@clementbera thanks for all the work so far. Before I continue trying to understand all the changes, what's the status with the tests? Any estimate on whether this is a fundamental issue or just about working through the various scattered small issues here and there?

File: src/som/interpreter/Method.java (hunk 4 above)
Comment (smarr) [reply src/som/interpreter/Method.java:82]: Not sure why I don't see these in the review, but we probably still want this to be `@CompilationFinal` for performance reasons.

File: src/som/interpreter/SArguments.java [hunk 6]
```diff
@@ -58,4 +91,139 @@ public static SImmutableArray getArgumentsWithoutReceiver(final Object[] argumen
...
+    if (VmSettings.ACTOR_ASYNC_STACK_TRACE_STRUCTURE) {
+      Object[] arguments = new Object[] {receiver, null};
+      entryLoad.loadShadowStackEntry(arguments, expression, frame, false);
+      return arguments;
+    } else {
+      return new Object[] {receiver};
+    }
+  }
+
+  public static Object[] getPlain1ArgumentWithReceiver(final Object receiver,
```
Comment (smarr) [inline src/som/interpreter/SArguments.java]: In Java, we got method overloading for better or for worse. So, I don't think we need to encode arity in method names here.

File: src/som/interpreter/SArguments.java [hunk 7]
```diff
@@ -58,4 +91,139 @@ public static SImmutableArray getArgumentsWithoutReceiver(final Object[] argumen
...
+    }
+  }
+
+  public static ShadowStackEntry instantiateTopShadowStackEntry(final Node expr) {
+    return ShadowStackEntry.createTop(expr);
+  }
+
+  public static ShadowStackEntry instantiateShadowStackEntry(final ShadowStackEntry previous,
+      final Node expr, final boolean async) {
+    if (async) {
```
Comment (smarr) [inline src/som/interpreter/SArguments.java:187]: Could you add the following? This is a compile time check that async is known and can be used for optimization. It's not a hint that this is the case, it will only result in an error if it isn't a constant value. Good for debugging of performance issues.

```suggestion
    CompilerAsserts.partialEvaluationConstant(async);
    if (async) {
```

File: src/som/interpreter/SArguments.java [hunk 8]
```diff
@@ -58,4 +91,139 @@ public static SImmutableArray getArgumentsWithoutReceiver(final Object[] argumen
...
+    Object maybeShadowStack = args[args.length - 1];
+    if (maybeShadowStack instanceof ShadowStackEntry) {
+      return (ShadowStackEntry) maybeShadowStack;
+    }
+    return null;
+  }
+
+  public static void setShadowStackEntry(final Object[] args,
+      final ShadowStackEntry entry) {
+    // assert args[args.length - 1] == null : "Assume shadow stack entry is not already set.";
```
Comment (smarr) [inline src/som/interpreter/SArguments.java]: Hm, why is this commented out?
Seems like a sensible assertion, or does it not hold? If so, in which case does it fail, should the assertion be removed or documented?

File: src/som/interpreter/nodes/dispatch/BackCacheCallNode.java [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    } else if (VmSettings.ACTOR_ASYNC_STACK_TRACE_STRUCTURE) {
+      SArguments.setShadowStackEntryWithCache(arguments, expression,
+          shadowStackEntryLoad, frame, false);
+    }
+    assert arguments[arguments.length - 1] != null
+        || (frame.getArguments()[frame.getArguments().length - 1] == null)
+        || !VmSettings.ACTOR_ASYNC_STACK_TRACE_STRUCTURE;
+  }
+
+  void uniqueCaller();
```
Comment (smarr) [inline src/som/interpreter/nodes/dispatch/BackCacheCallNode.java]: This and the next method seem to be mutating methods. Perhaps to be clear `set` or `make` would be a good prefix for them?

File: src/som/interpreter/nodes/dispatch/CachedDispatchNode.java [hunk 10]
```diff
@@ -2,19 +2,30 @@
...
 
-public final class CachedDispatchNode extends AbstractDispatchNode {
+public final class CachedDispatchNode extends AbstractDispatchNode
+    implements BackCacheCallNode {
 
-  @Child private DirectCallNode       cachedMethod;
-  @Child private AbstractDispatchNode nextInCache;
+  @Child private DirectCallNode         cachedMethod;
+  @Child private AbstractDispatchNode   nextInCache;
+  @CompilationFinal private boolean     uniqueCaller;
```
Comment (smarr) [inline src/som/interpreter/nodes/dispatch/CachedDispatchNode.java]: Same issue as in other node, you'll need an assumption to make this correct.

File: src/som/interpreter/nodes/dispatch/LexicallyBoundDispatchNode.java [hunk 11]
```diff
@@ -1,22 +1,33 @@
...
  * Private methods are special, they are linked unconditionally to the call site.
  * Thus, we don't need to check at the dispatch whether they apply or not.
  */
-public final class LexicallyBoundDispatchNode extends AbstractDispatchNode {
+public final class LexicallyBoundDispatchNode extends AbstractDispatchNode
+    implements BackCacheCallNode {
 
-  @Child private DirectCallNode cachedMethod;
+  @Child private DirectCallNode         cachedMethod;
+  @CompilationFinal private boolean     uniqueCaller;
```
Comment (smarr) [inline src/som/interpreter/nodes/dispatch/LexicallyBoundDispatchNode.java]: And all the other nodes: needs assumption

File: src/som/interpreter/Invokable.java (hunk 3 above)
Comment (clementbera) [reply src/som/interpreter/Invokable.java]: Done

File: src/som/interpreter/SArguments.java (hunk 6 above)
Comment (clementbera) [reply src/som/interpreter/SArguments.java]: So I changed to getPlainXArgumentWithReceiver with variable number of remaining parameters and made a single function instead of 3.

File: src/som/interpreter/SArguments.java (hunk 8 above)
Comment (clementbera) [reply src/som/interpreter/SArguments.java]: It should be uncommented. Done.

File: src/som/interpreter/SArguments.java (hunk 7 above)
Comment (clementbera) [reply src/som/interpreter/SArguments.java:187]: Done, it works.

File: src/som/interpreter/Method.java (hunk 4 above)
Comment (clementbera) [reply src/som/interpreter/Method.java:82]: I tried to fix the concern with invalidation (the 4 comments) in my latest commit.

Since the field uniqueCaller is used in Method only during relinking, it is not compiled by Graal and does not have to be compilation final. The fields uniqueCaller in CachedDispatchNode and LexicallyBoundDispatchNode have however to be. I created an assumption at creation time which is invalidated when the method makeMultipleCaller (was: multipleCaller) is executed.

Looks like some of the test code I have still runs that way.

**Comment by clementbera** [comment]
Before we added our change, there was already a try catch getting invalid assumption and replacing the node by the next in cache. This happened when guard.entryMatches raised the invalid assumption exception.

With new code this led to an infinite loop since the node was replaced with another node with the exact same guard.

I added a branch in the catch to check if the problem came from the guard or the stillUniqueCaller assumption. Depending on where it comes from, the problem is dealt with differently. In both cases, it leads in the end to a stable ast with a fast path.

**Comment by smarr** [comment]
@clementbera ok, thanks for the explanation!

File: src/som/interpreter/SArguments.java (hunk 8 above)
Comment (smarr) [reply src/som/interpreter/SArguments.java]: @ctrlpz please do not commit such a change. These assertions are there for a reason. It just cost me a lot of time that the assertion wasn't there.

**Comment by smarr** [comment]
This PR is out of date.

A new one may be opened based on the current branch derived from Carmen's version, which is based on Clement's code: https://github.com/smarr/SOMns/tree/ct/async-stack-traces
